src/_pytest/logging.py

```diff
@@ -613,7 +609,7 @@ def _runtest_for(self, item, when):
                 yield
 
     @contextmanager
-    def _runtest_for_main(self, item, when):
+    def _runtest_for_main(self, item, when: str) -> Generator[None, None, None]:
```

Comment:
@bluetech 
`item` not typed because of `catch_log_handler{,s}` being set/unset there.
Ignoring any refactors I've wondered if a `ItemWithLogging` type would make sense in general, which would have that attribute allowed then?  (I do not think it is worth it, was just wondering)

Reply:
Might work, but because the attribute is added and deleted, it can't be entirely correct, and would require casts as well, so probably doesn't make sense.

As for refactor, seems like a good case for an "item-level storage" mechanism https://github.com/pytest-dev/pytest/issues/3740. Such storage could be typed in some appropriate manner, probably.